lib/trace/tchannel.js

```diff
@@ -0,0 +1,163 @@
+        serviceName: 'ringpop',
+        endpoint: 'trace'
+    }
+};
+
+function TChannelTracer(ringpop, config, opts) {
+    opts.sink = _.extend(
+        {},
+        tchannelOptsDefaults.sink,
+        config.sinkOptsDefaults.tchannel,
+        opts.sink);
+    core.Tracer.call(this, ringpop, config, opts);
+}
+util.inherits(TChannelTracer, core.Tracer);
+
+
+TChannelTracer.prototype.validate = function validate() {
+    var sinkOpts = this.opts.sink;
+    return (
+        core.Tracer.prototype.validate.call(this) &&
+        sinkOpts &&
+        sinkOpts.type === 'tchannel' &&
+        sinkOpts.serviceName &&
+        sinkOpts.endpoint
+    );
+};
+
+TChannelTracer.prototype.connectSink = function connectSink(callback) {
+    if (callback) {
+        this.once('connectSink', callback);
+    }
+
+    var sinkOpts = this.opts.sink;
+    this.reqOpts = _.extend(
+        {
+            host: sinkOpts.hostPort,
+            timeout: 1000,
+            serviceName: 'ringpop',
+            hasNoParent: true,
+            trace: false,
+            retryLimit: 3,
+            headers: {
+                'as': 'raw',
+                'cn': 'ringpop'
+            }
+        },
+        _.omit(sinkOpts, 'type', 'hostPort')
+    );
+
+    // trigger early identification, minimize latency
+    this.ringpop.channel.waitForIdentified(
+        _.pick(this.reqOpts, 'host'),
+        this._remoteError.bind(this, 'connectSink'));
+};
+
+TChannelTracer.prototype.disconnectSink = function disconnectSink(callback) {
+    var self = this;
+    if (callback) {
+        this.once('disconnectSink', callback);
+    }
+
+    process.nextTick(function() {
+        self.disconnectSource();
+        self.emit('disconnectSink');   
+    });
+};
+
+TChannelTracer.prototype.send = function send(blob, callback) {
+    if (callback) {
+        this.once('send', callback);
+    }
+
+    if (typeof blob !== 'string') { // XXX also allow Buffer?
+        blob = JSON.stringify(blob);
+    }
+
+    this.ringpop.channel
+        .request(this.reqOpts)
+        .send(this.reqOpts.endpoint, blob, null, this._remoteError.bind(this, 'send'));     
+};
+
+TChannelTracer.prototype._remoteError = function _remoteError(event, err, resp) {
+    if (!err && resp && !resp.ok) {
+        err = new Error('TChannelTracer.' + event + ' received "notOk"');
+    }
+    if (err) {
+        if (typeof err !== 'object') {
+            err = new Error(err); // fallback for string
+        }
+        err.message = err.message +
+                ' [' + this.opts.hostPort + ' ' + this.reqOpts.endpoint + ']';
+        err.context = {
+            message: err.message,
+            reqOpts: this.reqOpts,
+            endpoint: this.reqOpts.endpoint,
+        };
+        this.ringpop.logger.warn(err.message);
+        this.emit(event, err);
+        return;
+    }
+    this.emit(event, null, resp);
+};
+
+TChannelTracer.prototype.matches = function matches(extEvent, opts) {
+    return (
```

Comment:
(discussed in person)

I argue for idempotence over the full src-dst pairing set { event, sinkType, sinkEndpoint, sinkURL, etc. }, since anything else allows duplicate sending to the same thing. Which is probably bad. And relying on a token means that one cli tool can't easily clean up after another.